src/ui/reducers/protocolMessages.ts

```diff
@@ -0,0 +1,74 @@
+import { EventMethods, EventParams } from "@recordreplay/protocol";
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+import cloneDeep from "lodash/cloneDeep";
+import { CommandRequest, CommandResponse } from "protocol/socket";
+import { UIState } from "ui/state";
+
+interface Recorded {
+  recordedAt: number;
+}
+
+type _ProtocolEvent<T extends EventMethods> = { method: T; params: EventParams<T> };
+export type ProtocolEvent = _ProtocolEvent<EventMethods>;
+export type RequestSummary = Omit<CommandRequest, "method"> & {
+  class: string;
+  method: string;
+  pending: boolean;
+  errored: boolean;
+  recordedAt: number;
+};
+
+export interface ProtocolMessagesState {
+  events: (ProtocolEvent & Recorded)[];
+  requests: RequestSummary[];
+  responses: (CommandResponse & Recorded)[];
+  errors: (CommandResponse & Recorded)[];
+}
+
+const protocolMessagesSlice = createSlice({
+  initialState: {
+    events: [],
+    requests: [],
+    responses: [],
+    errors: [],
+  } as ProtocolMessagesState,
+  name: "protocolMessages",
+  reducers: {
+    eventReceived(state, action: PayloadAction<ProtocolEvent & Recorded>) {
+      state.events.push(cloneDeep(action.payload));
+    },
+    responseReceived(state, action: PayloadAction<CommandResponse & Recorded>) {
+      state.responses.push(cloneDeep(action.payload));
+      const request = state.requests.find(r => r.id === action.payload.id);
+      if (request) {
+        request.pending = false;
+      }
+    },
+    errorReceived(state, action: PayloadAction<CommandResponse & Recorded>) {
+      state.errors.push(action.payload);
+      const request = state.requests.find(r => r.id === action.payload.id);
+      if (request) {
+        request.pending = false;
+        request.errored = true;
+      }
+    },
+    requestSent(state, action: PayloadAction<CommandRequest & Recorded>) {
+      state.requests.push({
+        ...action.payload,
+        class: action.payload.method.slice(0, action.payload.method.indexOf(".")),
```

Comment:
:hand: tbh I have a hard time reading the `.slice()` bits here :)

I _think_, from reading, that this is trying to take something like `Graphics.refreshSomething`, and split it?

Could you just do `const [ requestClass, requestMethod ] = action.payload.method.split(".")` and use those?

Reply:
haha, yes, thank you, that would have a been a much more intelligent way of phrasing it.